src/Lua.cs

```diff
@@ -496,7 +496,27 @@ public LuaType GetTable(LuaRegistry index)
         /// <param name="functionIndex"></param>
         /// <param name="n"></param>
         /// <returns>Returns the type of the pushed value. </returns>
-        public string GetUpValue(int functionIndex, int n) => NativeMethods.lua_getupvalue(_luaState, functionIndex, n);
+        public string GetUpValue(int functionIndex, int n) => MarshalString(NativeMethods.lua_getupvalue(_luaState, functionIndex, n));
+
+        /// <summary>
+        /// Converts a char* to a string, in a way that doesn't corrupt memory in specific circumstances
+        /// </summary>
+        /// <param name="ptr"></param>
+        /// <returns>The string</returns>
+        private string MarshalString(IntPtr ptr)
+        {
+            if (ptr == IntPtr.Zero)
+                return null;
+
+            if (Encoding == Encoding.ASCII)
+            {
+                return Marshal.PtrToStringAnsi(ptr);
+            }
+            else
+            {
+                return Marshal.PtrToStringUni(ptr);
```

Comment:
This should not be necessary. 
Return the string should already do the marshaling using `CharSet.Ansi`, the name would never be a unicode string.

Reply:
Good catch, I misunderstood what Encoding was for